File: source/downloads/code/LeetCode/Python/Super-Ugly-Number.py
import heapq

class Solution(object):
    def nthSuperUglyNumber(self, n, primes):
        """
        :type n: int
        :type primes: List[int]
        :rtype: int
        """

        uglies = [1]
        def gen(prime):
            for ugly in uglies:
                yield ugly * prime
        # merged: priority queue iterative
        merged = heapq.merge(*map(gen,primes))
         
        while len(uglies) < n:
            # current minimal value in merged, after uglies is appended.
            ugly = next(merged)
            if ugly != uglies[-1]:
                uglies.append(ugly)
        return uglies[-1]

if __name__ == "__main__":
    sol =Solution()
    sol.nthSuperUglyNumber(12,[2,3,5,7])

File: source/downloads/code/LeetCode/Java/Super-Ugly-Number.java
import java.util.Arrays;

public class Solution {
    public int nthSuperUglyNumber(int n, int[] primes) {
        int[] ugly = new int[n];
        //
        int[] idx = new int[primes.length];
        int[] val = new int[primes.length];

        Arrays.fill(val, 1);

        // next ugly value
        int next = 1;
        for(int i=0;i<n;i++){
            ugly[i] = next;

            next = Integer.MAX_VALUE;
            for( int j=0; j<primes.length;j++){
                if(val[j] == ugly[i])
                    val[j] = ugly[idx[j]++] * primes[j];
                next = Math.min(next, val[j]);
            }
        }

     return ugly[n-1]; 

    }
    public static void main(String arg[]){
    
        sol = Solution();
        sol.nthSuperUglyNumber(14, [2,7,13,19]);
    
    }
}
